### PR TITLE
feat(hydrator): enable controller

### PR DESCRIPTION
**Pull request by crenshaw-dev**
I'm breaking the hydrator into separate PRs. This is party to make them easier to review, partly because the commit history on the original PR is pretty messy at this point.

This PR is for the changes to the app controller to process the new sourceHydrator API + necessary tests.

I've added everyone who helped as a co-author on this PR.

### CODE REVIEW

File: manifests/namespace-install-with-hydrator/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+resources:
+  - ../cluster-install
```
Comment (ashutosh16) [inline manifests/namespace-install-with-hydrator/kustomization.yaml:2]: Is this correct? shouldn't it be namepsace-install

File: cmd/argocd/commands/app.go [hunk 2]
```diff
@@ -2644,21 +2647,23 @@ func waitOnApplicationStatus(ctx context.Context, acdClient argocdclient.Client,
 			}
 		}
 
+		hydrationFinished := app.Status.SourceHydrator.CurrentOperation != nil && app.Status.SourceHydrator.CurrentOperation.Phase == argoappv1.HydrateOperationPhaseHydrated && app.Status.SourceHydrator.CurrentOperation.SourceHydrator.DeepEquals(app.Status.SourceHydrator.LastSuccessfulOperation.SourceHydrator) && app.Status.SourceHydrator.CurrentOperation.DrySHA == app.Status.SourceHydrator.LastSuccessfulOperation.DrySHA
```
Comment (agaudreault) [inline cmd/argocd/commands/app.go:2650]: I would assume that `app.Status.SourceHydrator` can be null. Missing null check here? A unit test might be missing to catch this case.

File: controller/appcontroller.go [hunk 3]
```diff
@@ -2120,7 +2212,7 @@ func alreadyAttemptedSync(app *appv1.Application, commitSHA string, commitSHAsMS
 	} else {
 		// Ignore differences in target revision, since we already just verified commitSHAs are equal,
 		// and we do not want to trigger auto-sync due to things like HEAD != master
-		specSource := app.Spec.Source.DeepCopy()
+		specSource := app.Spec.GetSource()
```
Comment (agaudreault) [inline controller/appcontroller.go:2215]: DeepCopy was removed.
```suggestion
		specSource := app.Spec.GetSource().DeepCopy()
```

File: docs/operator-manual/argocd-cmd-params-cm.yaml [hunk 4]
```diff
@@ -9,13 +9,19 @@ data:
...
+
   # Redis server hostname and port (e.g. argocd-redis:6379)
   redis.server: "argocd-redis:6379"
   # Enable compression for data sent to Redis with the required compression algorithm. (default 'gzip')
   redis.compression: gzip
   # Redis database
   redis.db:
 
+  # Enables the alpha "manifest hydrator" feature. (default "false")
+  hydrator.enabled: "true"
```
Comment (agaudreault) [inline docs/operator-manual/argocd-cmd-params-cm.yaml:23]: All other parameters default value is the one set in the example

```suggestion
  hydrator.enabled: "false"
```

File: manifests/namespace-install-with-hydrator/kustomization.yaml (hunk 1 above)
Comment (agaudreault) [reply manifests/namespace-install-with-hydrator/kustomization.yaml:2]: +1

File: server/application/application_test.go [hunk 5]
```diff
@@ -109,7 +109,7 @@ func fakeResolveRevisionResponseHelm() *apiclient.ResolveRevisionResponse {
 
 func fakeRepoServerClient(isHelm bool) *mocks.RepoServerServiceClient {
 	mockRepoServiceClient := mocks.RepoServerServiceClient{}
-	mockRepoServiceClient.On("ListApps", mock.Anything, mock.Anything).Return(fakeAppList(), nil)
+	mockRepoServiceClient.On("GetProcessableApps", mock.Anything, mock.Anything).Return(fakeAppList(), nil)
```
Comment (agaudreault) [inline server/application/application_test.go:112]: This change seems unrelated to this PR 🤔

File: util/db/repository.go [hunk 6]
```diff
@@ -216,6 +216,20 @@ func (db *db) DeleteRepository(ctx context.Context, repoURL, project string) err
...
+	repository, err := db.repoWriteBackend().GetRepository(ctx, repoURL, project)
+	if err != nil {
+		return repository, fmt.Errorf("unable to get write repository %q: %w", repoURL, err)
+	}
+
+	// TODO: enrich with write credentials.
+	//if err := db.enrichCredsToRepo(ctx, repository); err != nil {
+	//	return repository, fmt.Errorf("unable to enrich write repository %q info with credentials: %w", repoURL, err)
+	//}
+
```
Comment (agaudreault) [inline util/db/repository.go:229]: Forgotten? Not a blocker, but create an issue if it is intended to implement later

File: controller/hydrator/hydrator.go [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+	defer argoio.Close(closer)
+	resp, err := commitService.CommitHydratedManifests(context.Background(), &manifestsRequest)
+	if err != nil {
+		return "", fmt.Errorf("failed to commit hydrated manifests: %w", err)
+	}
+	return resp.HydratedSha, nil
+}
+
+// appNeedsHydration answers if application needs manifests hydrated.
+func appNeedsHydration(app *appv1.Application, statusHydrateTimeout time.Duration, latestRevision string) string {
```
Comment (agaudreault) [inline controller/hydrator/hydrator.go:337]: Return `(needsHydration bool, reason string)` instead

File: cmd/argocd/commands/app.go (hunk 2 above)
Comment (crenshaw-dev) [reply cmd/argocd/commands/app.go:2650]: It can't be nil. I mimicked `SyncStatus`, which is a plain field on `ApplicationStatus` instead of a pointer.  `omitempty` is enabled, so if both fields inside `SourceHydrator` are `nil`, the field won't appear in the k8s API JSON.

File: controller/appcontroller.go (hunk 3 above)
Comment (crenshaw-dev) [reply controller/appcontroller.go:2215]: I think I did that on purpose, but I can't remember why.

Either way, this code is weird. Why are we emptying fields on temporary vars that we promptly discard?

File: util/db/repository.go (hunk 6 above)
Comment (crenshaw-dev) [reply util/db/repository.go:229]: Added docs, will check for an issue...

**Comment by crenshaw-dev** [comment]
@agaudreault @ashutosh16 @leoluz I addressed the easier comments, still working on the more involved ones.

**Comment by crenshaw-dev** [comment]
Parent PR was merged.
